File: rust/src/parser/statement.rs
// src/parser/statement.rs
use super::Parser;
use crate::ast::AST;
use crate::lexer::Token;
use crate::error::{CompileError, Result};

impl Parser {
    /// 文の解析
    pub(crate) fn parse_statement(&mut self) -> Result<AST> {
        self.handle_indentation()?;

        match self.peek_token() {
            // 改行を無視
            Some(Token::Newline) => {
                self.next_token();
                self.parse_statement()
            }

            // export (#) で始まる文
            Some(Token::Export) => {
                self.next_token();
                let expr = self.parse_expression()?;
                self.skip_newlines()?;
                Ok(AST::UnaryOp {
                    op: crate::ast::UnaryOperator::Export,
                    expr: Box::new(expr),
                })
            }

            // import (@) で始まる文
            Some(Token::Import) => {
                self.next_token();
                let expr = self.parse_expression()?;
                self.skip_newlines()?;
                Ok(AST::UnaryOp {
                    op: crate::ast::UnaryOperator::Import,
                    expr: Box::new(expr),
                })
            }

            // ブロックの開始
            Some(Token::LeftBracket | Token::LeftBrace | Token::LeftParen) => {
                let block = self.parse_block()?;
                self.skip_newlines()?;
                Ok(block)
            }

            // パターンマッチ構造
            Some(Token::Identifier(_)) if self.is_pattern_match() => {
                self.parse_pattern_match()
            }

            // その他の式
            _ => {
                let expr = self.parse_expression()?;
                self.skip_newlines()?;
                Ok(expr)
            }
        }
    }

    /// 改行をスキップ
    fn skip_newlines(&mut self) -> Result<()> {
        while let Some(Token::Newline) = self.peek_token() {
            self.next_token();
        }
        Ok(())
    }

    /// ブロックの解析
    fn parse_block(&mut self) -> Result<AST> {
        // 開始括弧の種類を記憶
        let start_token = self.next_token().unwrap();
        let end_token = match start_token {
            Token::LeftBracket => Token::RightBracket,
            Token::LeftBrace => Token::RightBrace,
            Token::LeftParen => Token::RightParen,
            _ => unreachable!(),
        };

        let mut statements = Vec::new();
        let mut expecting_separator = false;

        while let Some(token) = self.peek_token() {
            match token {
                // ブロック終了
                t if t == end_token => {
                    self.next_token();
                    break;
                }

                // セパレータ（改行または;）
                Token::Newline => {
                    self.next_token();
                    expecting_separator = false;
                }
                
                // 文の解析
                _ => {
                    if expecting_separator {
                        return Err(CompileError::InvalidSyntax(
                            "Expected separator between statements".to_string()
                        ));
                    }
                    statements.push(self.parse_statement()?);
                    expecting_separator = true;
                }
            }
        }

        Ok(AST::Block(statements))
    }

    /// パターンマッチ構造の判定
    fn is_pattern_match(&mut self) -> bool {
        let mut tokens = self.tokens.iter().take(3);
        match (tokens.next(), tokens.next(), tokens.next()) {
            (Some(Token::Identifier(_)), Some(Token::Define), Some(Token::Lambda)) => true,
            _ => false,
        }
    }

    /// パターンマッチ構造の解析
    fn parse_pattern_match(&mut self) -> Result<AST> {
        let mut patterns = Vec::new();
        let mut current_pattern = Vec::new();
        
        while let Some(token) = self.peek_token() {
            match token {
                // パターンとケースの区切り
                Token::Or => {
                    self.next_token();
                    if !current_pattern.is_empty() {
                        patterns.push(current_pattern);
                        current_pattern = Vec::new();
                    }
                }

                // ブロック終了
                Token::RightBracket | Token::RightBrace | Token::RightParen => {
                    if !current_pattern.is_empty() {
                        patterns.push(current_pattern);
                    }
                    break;
                }

                // パターンの一部
                _ => {
                    let expr = self.parse_expression()?;
                    current_pattern.push(expr);
                }
            }
        }

        Ok(AST::Match {
            value: Box::new(AST::Block(Vec::new())), // パターンマッチ対象
            cases: patterns.into_iter()
                .map(|pattern| {
                    let condition = if pattern.len() == 1 {
                        pattern[0].clone()
                    } else {
                        AST::Block(pattern)
                    };
                    (condition, AST::Block(Vec::new()))
                })
                .collect(),
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::lexer::Token;

    #[test]
    fn test_simple_statement() {
        let tokens = vec![
            Token::NumberLiteral(42.0),
            Token::Newline,
            Token::EOF,
        ];
        let mut parser = Parser::new(tokens);
        let ast = parser.parse_statement().unwrap();
        
        match ast {
            AST::Number(n) => assert_eq!(n, 42.0),
            _ => panic!("Expected number literal"),
        }
    }

    #[test]
    fn test_block_statement() {
        let tokens = vec![
            Token::LeftBracket,
            Token::NumberLiteral(1.0),
            Token::Newline,
            Token::NumberLiteral(2.0),
            Token::RightBracket,
            Token::EOF,
        ];
        let mut parser = Parser::new(tokens);
        let ast = parser.parse_statement().unwrap();
        
        match ast {
            AST::Block(statements) => assert_eq!(statements.len(), 2),
            _ => panic!("Expected block"),
        }
    }

    #[test]
    fn test_export_statement() {
        let tokens = vec![
            Token::Export,
            Token::Identifier("value".to_string()),
            Token::EOF,
        ];
        let mut parser = Parser::new(tokens);
        let ast = parser.parse_statement().unwrap();
        
        match ast {
            AST::UnaryOp { op: crate::ast::UnaryOperator::Export, .. } => {},
            _ => panic!("Expected export statement"),
        }
    }

    #[test]
    fn test_import_statement() {
        let tokens = vec![
            Token::Import,
            Token::Identifier("module".to_string()),
            Token::EOF,
        ];
        let mut parser = Parser::new(tokens);
        let ast = parser.parse_statement().unwrap();
        
        match ast {
            AST::UnaryOp { op: crate::ast::UnaryOperator::Import, .. } => {},
            _ => panic!("Expected import statement"),
        }
    }
}

File: javascript_AI/src/parser/statement.js
// src/parser/statement.js

const { TokenType } = require('../lexer/token');
const ExpressionParser = require('./expression');

class StatementParser {
    constructor(tokens = []) {
        this.tokens = tokens;
        this.current = 0;
        this.currentIndent = 0;
        this.expressionParser = new ExpressionParser(tokens);
    }

    parseStatement() {
        // インデントの処理
        this.handleIndentation();

        const token = this.peek();
        if (!token) return null;

        switch (token.type) {
            case TokenType.NEWLINE:
                this.advance();
                return this.parseStatement();

            case TokenType.EXPORT:
                return this.parseExportStatement();

            case TokenType.IMPORT:
                return this.parseImportStatement();

            case TokenType.LEFT_BRACKET:
            case TokenType.LEFT_BRACE:
            case TokenType.LEFT_PAREN:
                return this.parseBlockStatement();

            default:
                if (this.isPatternMatch()) {
                    return this.parsePatternMatch();
                }
                return this.parseExpressionStatement();
        }
    }

    handleIndentation() {
        while (this.match(TokenType.INDENT)) {
            this.currentIndent = this.previous().value;
        }
        while (this.match(TokenType.DEDENT)) {
            this.currentIndent--;
        }
    }

    parseExportStatement() {
        this.advance(); // skip #
        return {
            type: 'ExportStatement',
            expression: this.expressionParser.parseExpression()
        };
    }

    parseImportStatement() {
        this.advance(); // skip @
        return {
            type: 'ImportStatement',
            expression: this.expressionParser.parseExpression()
        };
    }

    parseBlockStatement() {
        const startToken = this.advance();
        const endToken = this.getMatchingEndToken(startToken.type);
        const statements = [];
        let expectingSeparator = false;

        while (this.peek() && !this.check(endToken)) {
            if (this.match(TokenType.NEWLINE)) {
                expectingSeparator = false;
                continue;
            }

            if (expectingSeparator) {
                throw new Error("Expected separator between statements");
            }

            statements.push(this.parseStatement());
            expectingSeparator = true;
        }

        this.advance(); // consume ending token

        return {
            type: 'BlockStatement',
            statements,
            style: this.getBlockStyle(startToken.type)
        };
    }

    parsePatternMatch() {
        const patterns = [];
        let currentPattern = [];

        while (this.peek() && !this.isBlockEnd()) {
            if (this.match(TokenType.OR)) {
                if (currentPattern.length > 0) {
                    patterns.push(currentPattern);
                    currentPattern = [];
                }
                continue;
            }

            const expr = this.expressionParser.parseExpression();
            currentPattern.push(expr);
            
            if (this.match(TokenType.DEFINE)) {
                const result = this.expressionParser.parseExpression();
                patterns.push({
                    pattern: currentPattern.length === 1 ? currentPattern[0] : currentPattern,
                    result
                });
                currentPattern = [];
            }
        }

        if (currentPattern.length > 0) {
            patterns.push({
                pattern: currentPattern.length === 1 ? currentPattern[0] : currentPattern,
                result: null // default case
            });
        }

        return {
            type: 'PatternMatch',
            cases: patterns
        };
    }

    parseExpressionStatement() {
        const expr = this.expressionParser.parseExpression();
        this.consumeNewline();
        return {
            type: 'ExpressionStatement',
            expression: expr
        };
    }

    // ユーティリティメソッド
    getMatchingEndToken(startToken) {
        const tokenMap = {
            [TokenType.LEFT_BRACKET]: TokenType.RIGHT_BRACKET,
            [TokenType.LEFT_BRACE]: TokenType.RIGHT_BRACE,
            [TokenType.LEFT_PAREN]: TokenType.RIGHT_PAREN
        };
        return tokenMap[startToken];
    }

    getBlockStyle(startToken) {
        const styleMap = {
            [TokenType.LEFT_BRACKET]: 'brackets',
            [TokenType.LEFT_BRACE]: 'braces',
            [TokenType.LEFT_PAREN]: 'parentheses'
        };
        return styleMap[startToken];
    }

    isPatternMatch() {
        let i = this.current;
        let hasDefine = false;
        let hasLambda = false;

        while (i < this.tokens.length) {
            const token = this.tokens[i];
            if (token.type === TokenType.DEFINE) hasDefine = true;
            if (token.type === TokenType.LAMBDA) hasLambda = true;
            if (hasDefine && hasLambda) return true;
            i++;
        }

        return false;
    }

    isBlockEnd() {
        return this.check(TokenType.RIGHT_BRACKET) ||
               this.check(TokenType.RIGHT_BRACE) ||
               this.check(TokenType.RIGHT_PAREN);
    }

    consumeNewline() {
        while (this.match(TokenType.NEWLINE)) {
            // consume all consecutive newlines
        }
    }

    // ExpressionParserと同じユーティリティメソッド
    peek() {
        return this.tokens[this.current] || null;
    }

    previous() {
        return this.tokens[this.current - 1];
    }

    advance() {
        if (!this.isAtEnd()) this.current++;
        return this.tokens[this.current - 1];
    }

    match(type) {
        if (this.check(type)) {
            this.advance();
            return true;
        }
        return false;
    }

    check(type) {
        if (this.isAtEnd()) return false;
        return this.peek().type === type;
    }

    isAtEnd() {
        return this.peek()?.type === TokenType.EOF;
    }
}

module.exports = StatementParser;
